core/field_textinput.js

```diff
@@ -365,12 +404,77 @@ Blockly.FieldTextInput.prototype.onHtmlInputChange_ = function(_e) {
+
+    // As I mentioned in the doValueUpdate_ function, the onHtmlInputChange
+    // handler deals with setting the display value.
+    // This makes sure we get the input "straight from the horses mouth"
+    // without any validators changing, or invalidating the input.
+
+    // This also means we have no need to override the doValueInvalid_
+    // function. A nice bonus.
+    this.displayValue_ = this.htmlInput_.value;
     this.forceRerender();
     Blockly.Events.setGroup(false);
   }
 };
 
+// Now you said that you did not want a solution that just renames text_.
+// Let me go ahead and speak to that for a moment.
+
+// Most of the problems with text_ have already been fixed prior to this PR.
+//   1) Setting the text_ property (for use in collapse text) was inefficient.
+//   This has been solved by moving people to the getText property.
+//   2) Moving peoples code to getText keeps it more organized, instead of
+//   having it split between doValueUpdate, doValueInvalid_ and render_.
+// As you know I also noted these facts in my first comment:
+// https://github.com/google/blockly/issues/2720#issuecomment-515688519
+
+// Now this next bit was my bad.
+// I decided that it was safest to avoid touching text_ to make sure the
+// code was as backwards compatible as possible, a decision I still stand by.
+// The problem with this is that in doing so I decided not to change how
+// getDisplayText_ generates its text.
+// *And in doing that I inadvertently imbued text_ with a power it did not
+// deserve, the power to change the display of the block.* I then leveraged
+// this in refactoring the text input field.
+
+// The changes demonstrated here take this ability away from the text_
+// property and give it to a property that actually deserves it, namely the
+// displayValue_ property.
+
+// The changes demonstrated here also solve the other remaining problem with
+// the text_ field, that being its ambiguity.
+// In the past text_ was used both for the display, and the collapsed text.
+// Now everything is more clear:
+//  * The value_ is the value held by the field, to be used in code generation.
+//  * getText() generates a string based on the value_, to be used on the
+//    collapsed block.
+//  * The displayValue is the value being displayed by the field.
+//  * getDisplayText() generates a string based on the displayValue_, to be
+//    used on the display of the block.
+// text_ is no longer used for anything.
+
+
+/* ---------------------------------------------------------------------------*/
+
+
+// I think that this is a good way to handle disempowering text_. But I do
+// realize one flaw of this strategy is that displayValue_ will often be a
+// duplicate of value_. I wanted to note that if instead you want to do:
+
+/*
+ * doValueUpdate(newValue) {
+ *   value_ = newValue;
+ *   tempDisplayValue_ = null;
+ * }
+ *
+ * getDisplayText() {
+ *   var text = tempDisplayValue || value;
+ * }
+ */
+
+// I would support this =)
```

Comment:
This was our original proposal. You may have found a better place to clear the tempDisplayValue_ in dropdownDispose_. If we set tempDisplayValue_ to null there it's a nice default of clearing it when the editor disappears. =)